src/lib/files.js

```diff
@@ -8,6 +8,10 @@ const ignore = [
 ]
 
 export async function filesToStreams (files) {
+  if (!Array.isArray(files)) {
```

Comment:
The last bug i fixed in file uploading was becuase the logic for dropping files didn't go through this function. 

We need to simplify the file upload path, so there is one common point at which we filter out files that should be ignored, and transform the input into the format expected by ipfs.add.

Is it possible to have the `selectDirectory` do less work, and just have it offer up the data as it would have been if the user had used the browser file input for adding a directory?

Reply:
Specifically, this PR https://github.com/ipfs-shipyard/ipfs-webui/pull/1011/files where the drag-n-drop code was duplicating some of the work that filesToStreams does, but was doing it slightly differently.

Reply:
> Is it possible to have the selectDirectory do less work, and just have it offer up the data as it would have been if the user had used the browser file input for adding a directory?

This function (`filesToStreams`) treats HTML5 File Streams, not Node.js Streams. I can't make `selectDirectory` return HTML5 File objects via a node.js fs function. At least as far as I know.

If you look at the code (https://github.com/ipfs-shipyard/ipfs-desktop/pull/927/files#diff-8c58d7bb18cd5b6faa7ff801849e00acR53), it basically just does native calls.

Reply:
So we need to figure out what the contract of `doFilesWrite` is, as it's dealing with too many edge cases.

We want one place in the code where the input is normalised and made ready for passing it to ipfs, and filtered for any files that shouldn't be there.

in the abstract, we want several specific, well named functions that handle the details of preparing the input from a specific scenario, and and they all convert the inputs into one shape, that is then passed on to one common function which always does the same thing.

We dont want to push many different types of input into one function and use detections at various points to adjust its  behaviour as that gets super fragile...

Reply:
@olizilla made some updates. Please take a look. Now we call filestostreams early on